# Exercises for Introduction to Python for Data Science

Week 08 - Pandas

Matthias Feurer and Andreas Bender  
2026-11-06

# Exercise 1

Create a pandas DataFrame that represents rapper information. The
DataFrame should contain the following data:

| Name           | Age | Genre   | Net_Worth_Millions | Albums_Released |
|----------------|-----|---------|--------------------|-----------------|
| Drake          | 37  | Hip-Hop | 250                | 8               |
| Kendrick Lamar | 36  | Hip-Hop | 75                 | 5               |
| J. Cole        | 39  | Hip-Hop | 60                 | 7               |
| Travis Scott   | 32  | Trap    | 80                 | 4               |
| Post Malone    | 28  | Hip-Hop | 45                 | 5               |

You should create this DataFrame using a dictionary where the keys are
the column names and the values are lists containing the data for each
column.

In [15]:
import pandas as pd

dic = {
    "Name": ["Drake", "Kendrick Lamar", "J.Cole", "Travis Scott", "Post Malone"],
    "Age": [37,46,39,32,28],
    "Genre": ["Hip-Hop", "Hip-Hop", "Hip-Hop", "Trap", "Hip-Hop"],
    "Net_Worth_Millions": [250, 75, 60, 80, 45],
    "Albums_Released": [8,5,7,4,5]
}

df = pd.DataFrame(dic)

# Ex 2

In [16]:
df[df["Genre"] == "Hip-Hop"]

Name  Age    Genre  Net_Worth_Millions  Albums_Released
0           Drake   37  Hip-Hop                 250                8
1  Kendrick Lamar   46  Hip-Hop                  75                5
2          J.Cole   39  Hip-Hop                  60                7
4     Post Malone   28  Hip-Hop                  45                5

In [23]:
print(df["Net_Worth_Millions"].mean(axis=0))

102.0


In [25]:
df[df["Albums_Released"] > 5]

Name  Age    Genre  Net_Worth_Millions  Albums_Released
0   Drake   37  Hip-Hop                 250                8
2  J.Cole   39  Hip-Hop                  60                7

In [28]:
df.sort_values("Net_Worth_Millions", ascending= False)

Name  Age    Genre  Net_Worth_Millions  Albums_Released
0           Drake   37  Hip-Hop                 250                8
3    Travis Scott   32     Trap                  80                4
1  Kendrick Lamar   46  Hip-Hop                  75                5
2          J.Cole   39  Hip-Hop                  60                7
4     Post Malone   28  Hip-Hop                  45                5

# Exercise 2

Using the rapper DataFrame from Exercise 1, perform the following
operations:

1.  Select all rappers who are Hip-Hop artists
2.  Find the average Net_Worth_Millions of all rappers
3.  Select rappers who have released more than 5 albums
4.  Sort the DataFrame by Net_Worth_Millions in descending order

Show the results of each operation.

# Exercise 3

Create a pandas Series that represents daily temperatures for a week.
The Series should have: - Index: Dates from January 1, 2023 to January
7, 2023 - Values: Temperatures \[7, 6, 9, 10, 8, 7, 8\] (in Celsius)

Then perform the following operations: 1. Calculate the mean temperature
2. Find the maximum and minimum temperatures 3. Calculate the
temperature range (max - min) 4. Convert temperatures to Fahrenheit
using the formula: (C \* 9/5) + 32

# Exercise 4

Create a DataFrame with missing values and demonstrate different methods
for handling them. Create the following DataFrame:

| City     | Jan_Temp | Feb_Temp | Mar_Temp |
|----------|----------|----------|----------|
| New York | 2        | NaN      | 8        |
| London   | 7        | 9        | NaN      |
| Berlin   | 3        | 5        | 10       |
| Paris    | 6        | 8        | 12       |

Then perform the following operations: 1. Check which values are missing
using `isnull()` 2. Count the number of missing values in each column 3.
Fill missing values with the mean of each column 4. Drop rows that
contain any missing values

# Exercise 5

Create a music streaming DataFrame and perform groupby operations.
Create the following data:

| Platform    | Artist         | Streams_Millions | Month    |
|-------------|----------------|------------------|----------|
| Spotify     | Drake          | 450              | January  |
| Apple Music | Kendrick Lamar | 320              | January  |
| Spotify     | J. Cole        | 280              | January  |
| Apple Music | Drake          | 380              | February |
| Spotify     | Travis Scott   | 290              | February |
| Apple Music | J. Cole        | 250              | February |

Then perform the following groupby operations:

1.  Group by Platform and calculate total streams for each platform
2.  Group by Artist and calculate average streams for each artist
3.  Group by Platform and Artist, then calculate the sum of streams
4.  Find the platform with the highest total streams

# Exercise 6

Create two DataFrames and perform merge operations. Create the following
data:

**Rappers DataFrame:**

| rapper_id | name           | city         |
|-----------|----------------|--------------|
| 1         | Drake          | Toronto      |
| 2         | Kendrick Lamar | Compton      |
| 3         | J. Cole        | Fayetteville |
| 4         | Travis Scott   | Houston      |

**Albums DataFrame:**

| album_id | rapper_id | album_name              | release_year |
|----------|-----------|-------------------------|--------------|
| 101      | 1         | Take Care               | 2011         |
| 102      | 2         | good kid, m.A.A.d city  | 2012         |
| 103      | 1         | Views                   | 2016         |
| 104      | 3         | 2014 Forest Hills Drive | 2014         |
| 105      | 5         | Astroworld              | 2018         |

Then perform the following merge operations:

1.  Inner merge (only rappers with albums)
2.  Left merge (all rappers, with album info where available)
3.  Right merge (all albums, with rapper info where available)
4.  Outer merge (all rappers and all albums)